src/os/bluestore/BlueStore.cc

```diff
@@ -8625,30 +8625,81 @@ int BlueStore::_move_ranges_destroy_src(TransContext *txc,
   dout(15) << __func__ << " " << c->cid << " "
            << srco->oid << " -> " << baseo->oid
            << dendl;
+  _dump_onode(srco, 30);
+  _dump_onode(baseo, 30);
+
+  baseo->exists = true;
+  _assign_nid(txc, baseo);
 
   int r = 0;
 
-  // Traverse move_info completely, move contents from src object
-  // to base object.
+  // fault in entire src; we'll need it for the delete anyway
+  srco->extent_map.fault_range(db, 0, OBJECT_MAX_SIZE);
+
+  // move all extents.  implicitly transfer all (referenced) blobs to
+  // the new object.
+  WriteContext wctx;
   for (unsigned i = 0; i < move_info.size(); ++i) {
-     uint64_t srcoff = move_info[i].get<0>();
-     uint64_t dstoff = move_info[i].get<1>();
-     uint64_t len = move_info[i].get<2>();
+    uint64_t srcoff = move_info[i].get<0>();
+    uint64_t dstoff = move_info[i].get<1>();
+    uint64_t length = move_info[i].get<2>();
+    dout(15) << __func__ << "   0x" << std::hex << srcoff << "~" << length
+	     << " to 0x" << dstoff << std::dec
+	     << dendl;
+    assert(dstoff >= srcoff);
 
-     dout(15) << __func__ << " " << c->cid << " " << srco->oid << " -> "
-              << baseo->oid << " from 0x" << std::hex << srcoff << "~" << len
-              << " to offset 0x" << dstoff << std::dec
-              << dendl;
+    baseo->extent_map.fault_range(db, dstoff, length);
+    baseo->extent_map.punch_hole(dstoff, length, &wctx.old_extents);
+
+    for (auto ep = srco->extent_map.seek_lextent(srcoff);
+	 ep != srco->extent_map.extent_map.end();
+	 ++ep) {
+      auto& e = *ep;
+      if (e.logical_offset >= srcoff + length) {
+	break;
+      }
+      dout(20) << __func__ << "  src " << e << dendl;
+      int skip_front, skip_back;
+      if (e.logical_offset < srcoff) {
+	skip_front = srcoff - e.logical_offset;
+      } else {
+	skip_front = 0;
+      }
+      if (e.logical_offset + e.length > srcoff + length) {
+	skip_back = e.logical_offset + e.length - (srcoff + length);
+      } else {
+	skip_back = 0;
+      }
+      Extent *ne = new Extent(
```

Comment:
Let's consider following disposition: 
e(10~90) -> blob(0~0x1000)
mv (10~90 - > 0x2000)
it results in
skip_front = 0, skip_back = 0
ne.logical_offset = 10 + 0 + 0x2000 - 10 = 0x2000
ne.blob_offset = 10 + 0 = 10
ne.length = 90 - 0 - 0 = 90
i.e.
new extent, located at alloc unit(AU) boundary(4K) has non zero blob offset. This violates our invariant that blobs are AU aligned and IMHO might have some negative impact on subsequent operations. Makes sense?